internal/util/ssa/managedfields.go

```diff
@@ -26,8 +26,75 @@ import (
 	"k8s.io/klog/v2"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	"sigs.k8s.io/cluster-api/internal/contract"
 )
 
+const classicManager = "manager"
+
+// DropManagedFields modifies the managedFields entries on the object that belong to "manager" (Operation=Update)
+// to drop ownership of the given paths if there is no field yet that is managed by `ssaManager`.
+//
+// If we want to be able to drop fields that were previously owned by the "manager" we have to ensure that
+// fields are not co-owned by "manager" and `ssaManager`. Otherwise, when we drop the fields with SSA
+// (i.e. `ssaManager`) the fields would remain as they are still owned by "manager".
+// The following code will do a one-time update on the managed fields.
+// We won't do this on subsequent reconciles. This case will be identified by checking if `ssaManager` owns any fields.
+// Dropping ownership in paths for existing "manager" entries (which could also be from other controllers) is safe,
+// as we assume that if other controllers are still writing fields on the object they will just do it again and thus
+// gain ownership again.
+func DropManagedFields(ctx context.Context, obj client.Object, ssaManager string, c client.Client, paths []contract.Path) error {
```

Comment:
```suggestion
func DropManagedFields(ctx context.Context, c client.Client, obj client.Object, ssaManager string, paths []contract.Path) error {
```
?

Reply:
Done.

Reply:
@ykakarap Please do the same for CleanUpManagedFieldsForSSAAdoption either in a separate follow-up or just part of the KCP PR